cpp/src/arrow/compute/api_vector.h

```diff
@@ -705,5 +737,52 @@ Result<std::shared_ptr<Array>> PairwiseDiff(const Array& array,
                                             bool check_overflow = false,
                                             ExecContext* ctx = NULLPTR);
 
+/// \brief Return the inverse permutation of the given indices.
+///
+/// For indices[i] = x, inverse_permutation[x] = i. And inverse_permutation[x] = null if x
+/// does not appear in the input indices. For indices[i] = x where x < 0 or x > max_index,
+/// it is ignored. If multiple indices point to the same value, the last one is used.
+///
+/// For example, with indices = [null, 0, 3, 2, 4, 1, 1], the inverse permutation is
+///   [1, 6, 3]                    if max_index = 2,
+///   [1, 6, 3, 2, 4, null, null]  if max_index = 6.
+///
+/// \param[in] indices array-like indices
+/// \param[in] options configures the max index and the output type
+/// \param[in] ctx the function execution context, optional
+/// \return the resulting inverse permutation
+///
+/// \since 19.0.0
+/// \note API not yet finalized
+ARROW_EXPORT
+Result<Datum> InversePermutation(
+    const Datum& indices,
+    const InversePermutationOptions& options = InversePermutationOptions::Defaults(),
+    ExecContext* ctx = NULLPTR);
+
+/// \brief Scatter the values into specified positions according to the indices.
+///
+/// For indices[i] = x, output[x] = values[i]. And output[x] = null if x does not appear
+/// in the input indices. For indices[i] = x where x < 0 or x > max_index, values[i]
+/// is ignored. If multiple indices point to the same value, the last one is used.
```

Comment:
```suggestion
/// in the input indices. For indices[i] = x where x < 0 or x > max_index, values[i]
/// is ignored. If multiple indices point to the same value, the last one is used.
```
Can you explain the point of the lenient behavior wrt. negative indices and the max_index option?
Is there a use case that this enables?

Reply:
No there isn't a particular case in my mind that `index < 0` or `index > max_index` can be very useful, so we can alternatively throw exceptions if we see such values, or even not detect them at all.

(And just in case you are curious about the option `max_index` itself, it is essential for cases that the input indices are "sparse", please see my other comment https://github.com/apache/arrow/pull/44394#discussion_r1823743699 for details of such usage.)

Thank you.

Reply:
OK. This is quite reasonable. I'll update.

Reply:
Done.